### ISSUE TITLE
Incorrect Timezone for Value Metrics

### ISSUE THREAD

**Issue by mislam** [laravel/nova-issues#2044]
- Laravel Version: 6.3.0
- Nova Version: 2.5.0
- PHP Version: 7.3.4
- Database Driver and Version: MySQL 5.7.25
- Operating System and Version: macOS 10.14.5
- Browser type and version: Google Chrome
- Reproduction Repository: None

### Description

The **Value Metrics** does not honor user's browser timezone set from `boot()` method in `NovaServiceProvider`. But it does that for the **Trend Metrics**.

The total number from the trend metrics (i.e. for 7 days) are not matching with the value metrics (i.e. for 7 days). The timezone still seems off. Note that, I set a custom timezone under the `boot()` method for `NovaServiceProvider` using `Nova::userTimezone`. My app's timezone is UTC, Nova's timezone is set to `Asia/Dhaka` (GMT+6) and my browser's timezone is `America/New_York`. I noticed, the value metrics' time-tracking starts from 7:00 AM instead of 12:00AM.

**config/app.php**
```
'timezone' => 'UTC',
```

**app/Providers/NovaServiceProvider.php**
```
public function boot()
{
    parent::boot();

    Nova::userTimezone(function (Request $request) {
        return 'Asia/Dhaka';
    });

    Nova::style('theme-overrides', public_path('css/nova.css'));
}
```

**app/Nova/Metrics/NewUsers.php**
```
public function ranges()
{
    return [
        'TODAY' => 'Today',
        ...
}
```
**Screenshot**

![Screenshot](https://i.imgur.com/aDg3qam.png)

The screenshot was taken on October 15, 2019 at 10:57 PM at New York's timezone. I checked in the database. The trend metrics shows the actual data, while the value metrics data isn't correct.

**Comment by jbrooksuk** [comment]
Hey @mislam, can you confirm that you have published Nova assets when updating?

**Comment by mislam** [comment]
@jbrooksuk Yes I did publish and cleared view cache as explained on the upgrade guide. I did the publish on my Local Valet and then pushed the changed to Git, then deployed on Forge. Do I need to manually SSH to the server and publish?

**Comment by jayjfletcher** [comment]
I am having the same issue since upgrading. Newly seeded db table, value metric shows 0 created.

Request payload being sent for the metric is.

Timestamps are UTC in the DB, but the query being run is using local timezone.

Payload being sent
![Screen Shot 2019-10-16 at 4 19 46 PM](https://user-images.githubusercontent.com/5743488/66955721-f15cd480-f030-11e9-949f-a095e129b269.png)

Query being run
![Screen Shot 2019-10-16 at 4 19 56 PM](https://user-images.githubusercontent.com/5743488/66955754-05083b00-f031-11e9-8110-db3fb9b3f030.png)

Timestamps in DB
![Screen Shot 2019-10-16 at 4 20 36 PM](https://user-images.githubusercontent.com/5743488/66955796-1c472880-f031-11e9-832a-39d39bce25b1.png)

**Comment by jayjfletcher** [comment]
Altering the value aggregate method to something like this worked for me...

```
/**
     * Return a value result showing the growth of a model over a given time frame.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Illuminate\Database\Eloquent\Builder|string  $model
     * @param  string  $function
     * @param  string|null  $column
     * @param  string|null  $dateColumn
     * @return \Laravel\Nova\Metrics\ValueResult
     */
    protected function aggregate($request, $model, $function, $column = null, $dateColumn = null)
    {
        $query = $model instanceof Builder ? $model : (new $model)->newQuery();

        $column = $column ?? $query->getModel()->getQualifiedKeyName();

        $timezone = $request->timezone;

        $timezoneShift = (function (Carbon $dateTime){
            return $dateTime->setTimezone(config('app.timezone'));
        });

        $previousRange = $this->previousRange($request->range, $timezone);
        $currentRange = $this->currentRange($request->range, $timezone);

        $previousValue = round(with(clone $query)->whereBetween(
            $dateColumn ?? $query->getModel()->getCreatedAtColumn(),
            array_map($timezoneShift,$previousRange)
        )->{$function}($column), $this->precision);

        return $this->result(
            round(with(clone $query)->whereBetween(
                $dateColumn ?? $query->getModel()->getCreatedAtColumn(),
                array_map($timezoneShift,$currentRange)
            )->{$function}($column), $this->precision)
        )->previous($previousValue);
    }
```

**Comment by ndberg** [comment]
For me its the other way around, the Value metric is correct while the Trend metric is incorrect.

@mislam why do you have different time zones in app & nova?

**Comment by demency** [comment]
I have userTimeZone setted to 'America/Santiago'. And my trend metrics doesn't works correctly:

The last value, the time is 20:15. (That's not correct)
![Captura de pantalla de 2019-10-30 23-05-33](https://user-images.githubusercontent.com/8731267/67912853-f90c9500-fb69-11e9-8f25-99eb7016303a.png)

The second value, the time is 23:05. (That's correct)
![Captura de pantalla de 2019-10-30 23-07-09](https://user-images.githubusercontent.com/8731267/67912886-16d9fa00-fb6a-11e9-9eac-412d8a65848e.png)

😕

**Comment by jbrooksuk** [comment]
Timezones are hard haha.

Okay, so I need to know exactly what's going on here. How are trends working in place and not another?

**Comment by jayjfletcher** [comment]
@jbrooksuk 

> Timezones are hard haha.
> 
> Okay, so I need to know exactly what's going on here. How are trends working in place and not another?

Trend's ```aggregate``` function is using Chronos::now(), without a timezone being passed to it. So it must be using the server's default, which could be different from the application timezone. (should this use Carbon and the framework's now helper? )

Value's ```aggregate``` function is using Carbon via the now helper, with the users time zone. 

Neither Trend nor Value are converting to the application's timezone before running the query. Which is what the db date columns will be saved in.

The code I shared above  for Value's aggregate sets the timezone to the application's timezone just before running the query.

**Comment by mislam** [comment]
@jayjfletcher I think you might be right. Although I don't know the insight of the code, but what I see in my app is that the trend metrics is using the server's default timezone (from config/app.php) instead of what's set in `NovaServiceProvider` (as I mentioned in this original issue). You should not question the end-developer why they are not using the same timezone for both app config and for Nova. Rather you should fix the aggregate function so that it respects Nova's timezone.

**Comment by jayjfletcher** [comment]
The db query itself should use the timezone in app.timezone because that's the timezone any date columns will be based on in the db. Saved dates won't be based on the nova user timezone. And using a timezone in the query that is different than what the columns are based on will throw off any queries. 

But anything outside of the actual query, yes respect whatever user timezone is set. This is were any conversions should take place. And looking at the code, I believe the user timezone is respected outside of the query.

If you want to test it out, take the two methods above, turn them into traits, and then use them on your respective metrics. I believe you'll get the results you're expecting.

**Comment by demency** [comment]
For some reasons, idk why 🤷‍♂️, my value metrics works correctly. Considering that if the timezone is changed using user timezone settings, the metrics will receive the corresponding parameter. Trend and Value metrics receive that parameter ("America/Santiago", "Asia/Dhaka", etc) as string. Specifically, Value metrics has all date respecting that parameter. In Trend metrics, instead, has an unexpected behavior. See the following code:

```php
    /**
     * Get the timezone offset for the user's timezone.
     *
     * @return int
     */
    public function offset()
    {
        if ($this->timezone) {
            return (new DateTime(Chronos::now()->format('Y-m-d H:i:s'), new DateTimeZone($this->timezone)))->getOffset() / 60 / 60;
        }

        return 0;
    }
```

In before code, Chronos::now() generate a date considering the server timezone to create a DateTime instance using the timezone parameter. In my case, the server is configured using "America/Santiago". if you make a request using the parameter as "America/Santiago", this function calc -3. Understanding that it should return a zero. Well, some change i made on this function in order to get correctly the offset.


```php
    /**
     * Get the timezone offset for the user's timezone.
     *
     * @return int
     */
    public function offset()
    {
        if ($this->timezone) {
            return round(Carbon::parse(now($this->timezone)->format('Y-m-d H:i:s'))->diffInSeconds(now($this->timezone)) / 60 / 60);
        }

        return 0;
    }
```

Chronos was replaced with **now** helper and timezone was added. **Round** method was added too because the diffInSeconds function returns a float. Maybe comparing using that method isn't the best way, but solve my issue.

**Comment by mislam** [comment]
Just found another discrepancy in the `currentRange` and  `previousRange` methods. Correct me if I'm wrong. See how the `now()` function uses the timezone. In my test, I was convinced to change the code to be as follows:

![Screenshot](https://i.imgur.com/4llTkBR.png)

**Comment by bolechen** [comment]
Have same confused here. i found when the mysql timezone is not  UTC, at `TrendDateExpression.php` line 63

```php
return (new DateTime(Chronos::now()->format('Y-m-d H:i:s'), new DateTimeZone($this->timezone)))->getOffset() / 60 / 60;
```

The `->getOffset()` will return wrong value.

**Comment by bolechen** [comment]
add this macro will tmp fix this:

```php
            TrendDateExpressionFactory::macro('mysql', function ($query, $column, $unit, $timezone) {
                return new MySqlTrendDateExpression($query, $column, $unit, 'UTC');
            });
```

**Comment by bolechen** [comment]
And i see laravel have an not document config `database.connections.mysql.timezone`, seems nova not used this.

**Comment by ifaniqbal** [comment]
In my case, trend metric jumps to the next day too soon. I set my timezone in Asia/Jakarta, which is UTC+7, and trend metric already showing trend for the next day at 5pm.
![image](https://user-images.githubusercontent.com/2200129/72606439-2be84d00-3951-11ea-87b3-8c5c57ab64d4.png)

**Comment by davidhemphill** [comment]
So I'm taking a look into this today 👋.

My understanding of the issue is that when Nova's metrics are calculated, they use the timezone of the user's browser. So if you have records that are in the future in UTC time relative to your local time, you will not see them in the aggregate results. I think this makes sense actually,  even though it's confusing. 

This confusing behavior was @taylorotwell's motivation for adding the `Nova::userTimezone(callback)`. The only problem is we don't use that when calculating metrics either. I think we probably should.

**Comment by ifaniqbal** [comment]
Trend metric still error with version 2.10.1.

**Comment by matthewjumpsoffbuildings** [comment]
Any news on this? Im still getting incorrect trend results for a nova app running in the Sydney timezone

**Comment by kayacekovic** [comment]
My DB is UTC and Laravel application is UTC+3.

This is the query a simple trend produce:

```sql
select date_format(`created_at` + INTERVAL 3 HOUR, '%Y-%m-%d') as date_result,
       count(`leads`.`id`)                                     as aggregate
FROM `leads`
WHERE `created_at` between '2020-02-28 00:00:00' and '2020-03-05 21:04:57'
  AND `leads`.`deleted_at` is null
group by date_format(`created_at` + INTERVAL 3 HOUR, '%Y-%m-%d')
order by `date_result` asc;
```

Dates are correct. Problem is with `WHERE created_at`...

So I've changed the query as following:

```sql
WHERE DATE_ADD(created_at, INTERVAL 3 HOUR) between '2020-02-28 00:00:00' and '2020-03-05 21:41:24'
```

And it worked!

I had to manipulate `src/Metrics/Trend.php` line 498.

```php
->whereBetween(\DB::raw('DATE_ADD('.$dateColumn.', INTERVAL 3 HOUR)'), [$startingDate, $endingDate])
```

**Comment by cwilby** [comment]
I think there is a redundant timezone conversion happening in the `aggregate` function in `Laravel\Nova\Metrics\Value`.

In most web applications, dates should be stored in a consistent timezone (usually UTC), and converted relatively from that consistent timezone, to another timezone, in the client application (e.g mobile app, SPA).  

In cases where timezones are based on user preference, my personal view is to null coalesce that value during conversion, defaulting to the detected timezone.

In the case of Nova, the web application requests resources from `GET /nova-api/{resource}s`, which responds with UTC dates.

![image](https://user-images.githubusercontent.com/13686317/79677869-5b099700-81aa-11ea-8c1b-0ee381813d1b.png)

These are then correctly converted and displayed in the UI.

![image](https://user-images.githubusercontent.com/13686317/79677897-999f5180-81aa-11ea-8575-77a8d32551cb.png)

However, in the case of querying for value metrics, the client does not play a role in converting dates based on the users preference or detected timezone.

Instead, `GET /nova-api/metrics/{metric}` performs all of the computation, which is based on the server, which should be using the consistent timezone (usually UTC).

That endpoint calls the metric's `calculate` method through polymorphism, which for a Value metric looks something like this:

```
    public function calculate(NovaRequest $request)
    {
        return $this->count($request, User::class);
    }
```

`count` calls `aggregate`

```php
    public function count($request, $model, $column = null, $dateColumn = null)
    {
        return $this->aggregate($request, $model, 'count', $column, $dateColumn);
    }
```

and `aggregate` calls `previousRange` and `currentRange` to get the ranges used to query dates in the consistent timezone (usually UTC).

![image](https://user-images.githubusercontent.com/13686317/79678032-e7688980-81ab-11ea-8e29-0c8dadd70409.png)

**Comment by cwilby** [comment]
I don't think it's necessary to do timezone conversion for value metrics.  

All events that happened in the last 24 hours, happened in the same moment for all timezones.

Two people in two different timezones log in to the application at the same moment.

What should they see?

**Comment by cwilby** [comment]
In Laravel apps, dates are stored in the timezone specified in `app.timezone`.  The default is UTC. 

In most cases, Nova requests resources from `GET /nova-api/{resource}s` and receives dates in that timezone.

![image](https://user-images.githubusercontent.com/13686317/79677869-5b099700-81aa-11ea-8c1b-0ee381813d1b.png)

It then converts these dates to the users local time.  (In my case, `America/Los_Angeles`).

![image](https://user-images.githubusercontent.com/13686317/79677897-999f5180-81aa-11ea-8575-77a8d32551cb.png)

When querying value metrics, Nova requests a count from `GET /nova-api/metrics/{metric}?timezone=America%2FLos_Angeles`.

For a value metric using `count`, Nova uses the `previousRange` and `currentRange` helper to calculate inputs for the two `whereBetween` calls made in the `aggregate` function.

The problem is, these dates are in `America/Los_Angeles`, and are about to query against dates stored in `app.timezone`.

![image](https://user-images.githubusercontent.com/13686317/79678032-e7688980-81ab-11ea-8e29-0c8dadd70409.png)

If I'm in -0800, I only see 16 hours worth of data.
If I'm in +0200, I only see 22 hours of data. 

I could set the Nova timezone to UTC, but then none of the dates would be formatted correctly.

Hope I've shared my thoughts clearly enough - [dates are hard](https://www.youtube.com/watch?v=-5wpm-gesOY).

**Comment by cwilby** [comment]
I don't think timezone conversion is necessary for value metrics - (willing to be wrong).

When querying value metrics, Nova requests a count from `GET /nova-api/metrics/{metric}?timezone=America%2FLos_Angeles`.

For a value metric using the `count` method, Nova needs to get the start/end dates for the current and previous range.

The problem - the dates used for these ranges will be queried against dates stored in the timezone set by `config('app.timezone')`.

Removing abstraction, this are the queries that are called.

`database`
```
users: [
  { id: 1, created_at: '2020-04-19 03:09:40' }
]
```

This is the query:

```
$previousCount = with(clone $query)->whereBetween(
  'created_at',
  [
    "2020-02-18 20:17:04",
    "2020-03-19 20:17:04"
  ]
);

$currentCount = with(clone $query)->whereBetween(
  'created_at',
  [
    "2020-03-19 20:17:10",
    "2020-04-18 20:17:10"
  ]
);
```

Hope I've shared my thoughts clearly enough - [dates are hard](https://www.youtube.com/watch?v=-5wpm-gesOY).

**Comment by cwilby** [comment]
Setting `Nova::userTimezone` doesn't solve the issue, as Nova attaches the browsers detected timezone when querying for value metrics.

**Comment by cwilby** [comment]
When I remove all timezone conversions from `Laravel\Nova\Metrics\Value`, everything works for the test cases I provided.

**Comment by ifaniqbal** [comment]
Can you share the patch for it? I want to try it. Thanks

**Comment by cwilby** [comment]
@davidhemphill - wanted to make sure I address this point also.

> My understanding of the issue is that when Nova's metrics are calculated, they use the timezone of the user's browser. So if you have records that are in the future in UTC time relative to your local time, you will not see them in the aggregate results. I think this makes sense actually, even though it's confusing.

If I have records in the future in UTC time, I would not see them in the aggregate results as long as the range value calculations and database values are using the same timezone.

**Comment by cwilby** [comment]
@ifaniqbal https://gist.github.com/cwilby/6a63eb44bf2e89759410b640c757035f

**Comment by cwilby** [comment]
Hey @ifaniqbal, did this patch resolve your issue?

**Comment by ifaniqbal** [comment]
Sorry, I just remembered that the problem I experienced was in the trend
metric, not the value metric.

**Comment by cwilby** [comment]
@ifaniqbal I'll try refactoring the same aggregate function in the `Trend` class.    

If you want to try the patch yourself, the goal is to test what happens by removing timezone conversion logic.   It's mostly doing find all for "$timezone", removing code, and checking results in Nova.

**Comment by smartens80** [comment]
Any progress with this by any chance?

**Comment by cwilby** [comment]
@smartens80 I started a private fork with timezone conversions removed from metric classes and it works well.  

Here's `Laravel\Nova\Metrics\Value` refactored without timezone conversion

https://gist.github.com/cwilby/6a63eb44bf2e89759410b640c757035f

**Comment by GDanielRG** [comment]
Such a shame that this is part of the core feature of a paid product and 3 years later that feature cannot be used properly. I just wasted 3 days of my time trying to figure out why my metrics where not working blaming the configuration and coding on me, to find out its a problem with the product itself.

**Comment by oceanapplications** [comment]
This is still an ongoing problem. My server time is America/New_York, config.app.timezone is America/New_York, my Nova boot method is setting the timezone to America/New_York. 

The query generated is:
``
select
  date_format(`created_at`, '%Y-%m-%d') as date_result,
  count(`orders`.`id`) as aggregate
from
  `orders`
where
 `created_at` between '2020-10-13 00:00:00'
  and '2020-10-19 12:35:46'
  and `orders`.`deleted_at` is null
group by
  date_format(`created_at` - INTERVAL 4 HOUR, '%Y-%m-%d')
order by
  `date_result` asc
``

The problem is the INTERVAL 4 HOUR being added when the data is already timezone adjusted. I end up with data from 2020-10-12 and that kills the charts.

**Comment by adebonis** [comment]
Still an issue - were building an important Covid Vaccine signup app and really could use this fixed

**Comment by crynobone** [comment]
It's extremely hard to follow the replies above:

* Original issues are more than 1 year old.
* Some reply with `INTERVAL` SQL, which only exists on Trend metric, and not Value Metric as the title. Both should be debugged and reproduce separately.

I can look into it tomorrow but would be useful if each person submits a separate issue with reproducing examples so we can understand each use case.

**Comment by crynobone** [comment]
This has been fixed and released in v3.21.0
